STYLE.md

```diff
@@ -0,0 +1,432 @@
+```
+
+#### End-of-Line Comments
+
+Separate end-of-line comments from the code with 2 spaces. Align
+comments for data type definitions. Some examples:
+
+```haskell
+data Parser =
+  Parser
+    !Int         -- ^ Current position
+    !ByteString  -- ^ Remaining input
+
+foo :: Int -> Int
+foo n = salt * 32 + 9
+  where
+    salt = 453645243  -- Magic hash salt.
+```
+
+#### Links
+
+Use in-line links economically.  You are encouraged to add links for
+API names. It is not necessary to add links for all API names in a
+Haddock comment. We therefore recommend adding a link to an API name
+if:
+
+* The user might actually want to click on it for more information (in
+  your opinion), and
+
+* Only for the first occurrence of each API name in the comment (don't
+  bother repeating a link)
+
+Top-down guideline
+------------------
+
+### LANGUAGE extensions section
+
+Write each `LANGUAGE` pragma on its own line, sort them alphabetically and
+align by max width among them.
+
+```haskell
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE Rank2Types       #-}
+{-# LANGUAGE TemplateHaskell  #-}
+```
+
+### Module name
+
+Use singular when naming modules (e.g. use `Data.Map` and
+`Data.ByteString.Internal` instead of `Data.Maps` and
+`Data.ByteString.Internals`). Sometimes it's acceptable to use plural
+(e. g. `Types`, `Instances`).
+
+### Export Lists
+
+Format export lists as follows:
+
+```haskell
+module Data.Set
+  ( Set
+  , empty
+  , singleton
+  , member
+  ) where
+```
+
+Some clarifications:
+
+1. Use 2 spaces indentation for export list.
+2. You can split export list into sections or just write all as single section.
+3. It is strongly adviced to sort each section alpabetically. However,
+   classes, data types and type aliases should be written before
+   functions.
+
+### Imports
+
+Imports should be grouped in the following order:
+
+0. Implicit import of custom prelude (for example [`cardano-prelude`](https://github.com/input-output-hk/cardano-prelude)) if used.
+1. External modules.
+2. External IOHK modules.
+3. Local library modules.
+4. Local test modules
+
+* Put a blank line between each group of imports.
+
+* The imports in each group should be sorted alphabetically, by module name.
+
+* Always use explicit import lists or `qualified` imports.
+Try to use `qualified` imports only if import list is
+big enough or there are conflicts in names. This makes the code more robust
+against changes in these libraries. Exceptions:
+  1. Any custom prelude (e.g. [`cardano-prelude`](https://github.com/input-output-hk/cardano-prelude))
+  2. Modules that only re-export stuff from other modules
+
+If `import` is unqualified then put _11 spaces_ between `import` keyword and
+module name (i.e. length of `qualified` + 2).
+
+Unqualified types (i.e. `Map` vs. `M.Map`) look pretty good and not so ugly.
+Prefer two-line imports for such standard containers.
+
+```haskell
+import           Data.Map (Map)
+import qualified Data.Map as M hiding (Map)
```

Comment:
Is the `hiding (Map)` bit adding value here?

Reply:
I would argue that it isn't. It's quite obvious from the unqualified `Data.Map` import that you're using `Map` unqualified so I see no reason to hide it from the qualified `Data.Map` import.

Reply:
Well actually... perhaps since it prevents you from using `M.Map`, that could add some value.

Reply:
I think so in terms in readability. Prefixing a `Map` related function with `M.` makes sense but having `M.Map` in  a type signature doesn't add value to me. It adds noise.

Reply:
I think there are two separate questions:

1. lack of merit of using `M.Map`
2. enforcement of #1 through `hiding` -- which has an extra bookkeeping cost attached.

It's easy to agree on 1, but the 2 requires us to recognise the cost and balance it against the benefit of turning the discouragement into prohibition.

Reply:
> I think there are two separate questions:
> 
> 1. lack of merit of using `M.Map`
> 2. enforcement of #1 through `hiding` -- which has an extra bookkeeping cost attached.
> 
> It's easy to agree on 1, but the 2 requires us to recognise the cost and balance it against the benefit of turning the discouragement into prohibition.

Yeah, I don't think I'd enforce `hiding` the symbols utilized in the unqualified import. Although, perhaps, it's a valid option if the dev wishes to do so. I'm indifferent on this point.

Reply:
@deepfire I agree we don't need to hide `Map`.

Reply:
I would say that for:
```
import qualified Data.Map as M
```
`M` is not a great qualified name. I think `Map` would be a far better qualified name.

Reply:
I would agree with @erikd., despite having (ab)used `M` for `Data.Map` a lot in the past.